framework/patch.go

```diff
@@ -0,0 +1,42 @@
+package framework
+
+type patchType string
+
+const (
+	patchCreate patchType = "create"
+	patchDelete patchType = "delete"
+	patchUpdate patchType = "update"
+)
+
+// Patch is a set of information required in order to reconcile to the desired
+// state. Patch is split into three parts: create, delete and update. The parts
+// are passed as arguments to Resource's Create, Delete and Update functions
+// respectively. Patch is guaranteed to be applied in that order (i.e. create,
+// update, delete).
+type Patch struct {
+	data map[patchType]interface{}
+}
+
+func NewPatch() *Patch {
+	return &Patch{
+		data: make(map[patchType]interface{}, 3),
+	}
+}
+
+func (p *Patch) getCreate() (interface{}, bool) {
+	create, ok := p.data[patchCreate]
+	return create, ok
+}
+
+func (p *Patch) getDelete() (interface{}, bool) {
+	delete, ok := p.data[patchDelete]
+	return delete, ok
+}
+func (p *Patch) getUpdate() (interface{}, bool) {
+	update, ok := p.data[patchUpdate]
+	return update, ok
+}
+
+func (p *Patch) SetCreate(create interface{}) { p.data[patchCreate] = create }
+func (p *Patch) SetDelete(delete interface{}) { p.data[patchDelete] = delete }
+func (p *Patch) SetUpdate(update interface{}) { p.data[patchUpdate] = update }
```

Comment:
Do we want to align the naming with the resource and call the `Set*Patch`. I would like to do this. When you are implementing a resource and you are doing `p.SetDelete` within `r.NewDeletePatch` my heart cries.

Reply:
Does your heart cry when you do `p.SetCreatePatch` and then `p.SetUpdatePatch` within `r.NewUpdatePatch`? What about calling `p.SetUpdatePatch` inside `r.NewDeletePatch`? That also can happen. Those are separate things. One is about creating `*Patch` struct and second about filling the `*Patch` struct with actual data. I don't like repetition so `Patch.SetCreatePatch` doesn't really looks like a good name for me, I'd rather call it `Patch.SetCreateData` if you want some common suffix.

Tell me what you think. I change it to w/e you want, even to `Patch.SetCreatePatch`.

Reply:
> Does your heart cry when you do `p.SetCreatePatch` and then `p.SetUpdatePatch` within `r.NewUpdatePatch`?

No, this would make sense to me. This was what I asked for. 

> What about calling `p.SetUpdatePatch` inside `r.NewDeletePatch`?

I wondered about this initially but it makes sense to me. Most operators will just not do it because they do not need to, but being able to apply updates on delete events makes total sense even though I cannot think about a real use case from the top of my head right now. 

> I'd rather call it `Patch.SetCreateData`

Then lets rather stick to the `State` term and lets call it `Set*State`. What about this? Calling `p.SetCreateState` within `r.NewDeletePatch` is reasonable for me.

Reply:
Then also `Apply*State`?

Reply:
That would probably be the consequence.

Reply:
I think I got the right name :D I think patch isn't about state, patch is about change. So maybe `Set*Change` and `Apply*Change`?

That would be a nice interface IMO:

```
Patch {
   getCreateChange() : interface{}, bool
   getDeleteChange() : interface{}, bool
   getUpdateChange() : interface{}, bool

   SetCreateChange(create interface{})
   SetDeleteChange(delete interface{})
   SetUpdateChange(update interface{})
}
```

```
Resource {
   GetCurrentState(ctx context.Context, obj interface{}) : interface{}, error
   GetDesiredState(ctx context.Context, obj interface{}) : interface{}, error

   NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) : *Patch, error
   NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) : *Patch, error

   ApplyCreateChange(ctx context.Context, obj, createPatch interface{}) : error
   ApplyDeleteChange(ctx context.Context, obj, deletePatch interface{}) : error
   ApplyUpdateChange(ctx context.Context, obj, updatePatch interface{}) : error
}
```

This nicely describes the flow: a patch is created from the diff of current and desired state. And then the changes described by the patch are applied.

Reply:
I do not like the naming of `Change` but I like the bigger picture. Lets go with that then.

Reply:
Done.